src/main/scala/com/databricks/labs/overwatch/Optimizer.scala

```diff
@@ -31,22 +31,68 @@ object Optimizer extends SparkSessionWrapper{
       .orderBy('Pipeline_SnapTS.desc)
   }
 
+  /**
+   * This method is used to get the API URL from config table.
+   *
+   * @param orgID      The organization ID as a string.
+   * @param dataTarget The data target object.
+   * @return The API URL as a string.
+   *
+   *         The method works as follows:
+   *         1. It retrieves the ETL storage prefix from the data target and trims any leading or trailing spaces.
+   *         2. It constructs the report location by appending "report/configTable" to the last 12 characters of the ETL storage prefix.
+   *         3. It defines a window function that partitions by workspace_id and orders by snapTS in descending order.
+   *         4. It reads the data from the report location and selects the workspace_id, api_url, and snapTS columns.
+   *         5. It filters the data to only include rows where the workspace_id matches the provided organization ID.
+   *         6. It adds a new column "rnk" which is the rank of each row within its partition, as defined by the window function.
+   *         7. It filters the data to only include rows where "rnk" is 1, i.e., the row with the latest snapTS for each workspace_id.
+   *         8. It returns the api_url of the first row in the resulting DataFrame.
+   */
+  private def getApiURL(orgID: String,dataTarget: DataTarget): String = {
+    val etlStoragePrefix = dataTarget.etlDataPathPrefix.get.trim
+    val reportLocation = etlStoragePrefix.substring(0,etlStoragePrefix.length-12)+"report/configTable" //Removing global prefix and adding report location
+    if(Helpers.pathExists(reportLocation)) { //If the config table is found at the specified location, proceed with legacy optimization.
+      val latestRunWindow = Window.partitionBy('workspace_id).orderBy('snapTS.desc)
+      spark.read.load(reportLocation).select("workspace_id", "api_url", "snapTS")
+        .filter('workspace_id === orgID)
+        .withColumn("rnk", rank.over(latestRunWindow))
+        .filter('rnk === 1)
+        .first().getString(1)
+    } else {  // If the config table is not found at the specified location, throw an exception.
+      throw new BadConfigException(s"Config table not found at $reportLocation")
+    }
+
+  }
+
+
   /**
    * Derive the workspace from the config supplied in the last run
    * @param overwatchETLDB name of the Overwtch ETL database
    * @return workspace object
    */
   private def getLatestWorkspace(overwatchETLDB: String,orgId: String): Workspace = {
-    val params = getLatestSuccessState(overwatchETLDB,orgId)
+   val latestSuccessState = getLatestSuccessState(overwatchETLDB,orgId).limit(1)
+   val orgID = latestSuccessState.select("organization_id").collect().map(x => x(0).toString)
+    val params = latestSuccessState
       .selectExpr("inputConfig.*")
       .as[OverwatchParams]
       .first
 
     val args = JsonUtils.objToJson(params).compactString
```

Comment:
these two calls to `JsonUtils.objToJson` seem redundant.  consider instead:

```scala
val args = JsonUtils.objToJson(params)
logger.log(Level.INFO, s"ARGS: Identified config string is: \n${args.prettyString}")
try {
  val apiURL = getApiURL( orgID(0), params.dataTarget)
```
. . .
```scala
  Initializer( args.compactString, apiURL, orgID)
```
. . . 

Then you already have the `Option` types that you need, right?

Reply:
implemented